### PR TITLE
Change how manure handler configurations are handled

### LINKED ISSUE #1127: Hard-coded default values for manure handler config
The `ManureHandlerConfig` and `DefaultManureHandlerConfigFactory` hardcode some values when initializing default manure handlers. 
```python
@dataclass
class ManureHandlerConfig:
    """Class for storing the configuration of a manure handler.

    Attributes
    ----------
    cleaning_water_use_rate : float
        Amount of cleaning water used per animal per day, L.
    minutes_per_cleaning : int
        Number of minutes needed per animal per cleaning, minutes.
    cleanings_per_day : int
        Number of cleanings per day.
    daily_tillage_frequency : int
        Number of times per day that compost bedding is tilled.
    cleaning_water_recycle_fraction : float
        Fraction of cleaning water that is from recycled (not fresh) water sources.
    """

    cleaning_water_use_rate: float = 0.0
    minutes_per_cleaning: int = 8
    cleanings_per_day: int = 2
    daily_tillage_frequency: int = 0
    cleaning_water_recycle_fraction: float = 0.0


class DefaultManureHandlerConfigFactory:
    """Class for creating default manure handler configurations."""

    FLUSH_SYSTEM_CONFIG = ManureHandlerConfig(
        cleaning_water_use_rate=757.0,
        cleaning_water_recycle_fraction=0.80,
    )
    MANUAL_SCRAPING_CONFIG = ManureHandlerConfig(
        cleaning_water_use_rate=10.0,
        cleaning_water_recycle_fraction=0.10,
    )
    ALLEY_SCRAPER_CONFIG = ManureHandlerConfig(
        cleaning_water_use_rate=10.0,
        cleaning_water_recycle_fraction=0.10,
    )
    TILLAGE_CONFIG = ManureHandlerConfig(
        daily_tillage_frequency=1,
    )
    HARROWING_CONFIG = ManureHandlerConfig(
        daily_tillage_frequency=0,
    )
```
## Steps to Reproduce

1.
2.
3.
4.

### PR DESCRIPTION
**Pull request by Isgotkowitz**
This issue closes #1127, removing the hard coded defaults for the manure handler configuration options. Additionally, it preserves the high level of configurability for manure handlers but adds more error handling to prevent users from using non-existent manure handlers.

I am pretty confident that similar changes could be made with respect to the bedding, separator, and treatment configurations (but I have not verified this yet!). This ties into issue #1118.

### Old behavior
What was possible with the old system:
```json
 "manure_management_scenarios": [
    {
      "scenario_id": 0,
      "manure_handler": "flush system",
      -- snip -- 
    },
    {
      "scenario_id": 1,
      "manure_handler": "more efficient flush system",
      -- snip --
    }
  ]
-- snip -- 
  "manure_handler_configs": [
    {
      "manure_handler_type": "flush system",
      "cleaning_water_use_rate": 757.0,
      -- snip --
    },
    {
      "manure_handler_type": "more efficient flush system",
      "cleaning_water_use_rate": 750.0,
      -- snip --
    }
  ]
```

### New behavior
How the above would be specified now:
```json
 "manure_management_scenarios": [
    {
      "scenario_id": 0,
      "manure_handler": "flush system",
      -- snip -- 
    },
    {
      "scenario_id": 1,
      "manure_handler": "more efficient flush system",
      -- snip --
    }
  ]
-- snip -- 
  "manure_handler_configs": [
    {
      "name": "flush system",
      "manure_handler_type": "flush system",
      "cleaning_water_use_rate": 757.0,
      -- snip --
    },
    {
      "name": "more efficient flush system",
      "manure_handler_type": "flush system",
      "cleaning_water_use_rate": 750.0,
      -- snip --
    }
  ]
```

### So what's the difference?
By introducing the "name" field on top of "manure_handler_type" in "manure_handler_configs", variations of the same system can be specified. For example, this means that any number of custom flush systems could be added to the input, but RuFaS will still be able to categorize and group them as flush systems.

The new system is also much less tolerant of users incorrectly writing their manure management input file. There are three cases that will crash the simulation:
1. "manure_handler_type" is not correctly specified to be one of the following: "flush system", "manual scraping", "alley scraper", "tillage" or "harrowing".
2. There is more than one manure handler configuration with the same name.
3. If the "manure_handler" value in a manure management scenario does not match any of the names in the specified in "manure_handler_configs".

Closes #1127

### CODE REVIEW

File: input/metadata/properties/default.json [hunk 1]
```diff
@@ -3462,8 +3462,7 @@
         },
         "manure_handler": {
           "type": "string",
-          "description": "Manure Handling Method -- Method for cleaning barn alleyways.",
-          "pattern": "^(flush system|alley scraper|manual scraping|tillage|harrowing)$"
+          "description": "Manure Handling Method -- Method for cleaning barn alleyways."
```
Comment (KFosterReed) [inline input/metadata/properties/default.json]: Does it make sense to make it clear in this description that the entry here needs to correspond to the "name" in the manure handler config? 

Perhaps something like: "Name of the Manure Handling Method -- Method for cleaning barn alleyways. Must match 'name' attribute of desired Manure Handler Config"

**Review by KFosterReed — APPROVED** [review]
Nice work Ed! 

I made one suggestion about the metadata description for the manure handler name in the manure scenario definition but other than that, this works as expected! I was able to run all the different manure handler types and specify multiple handlers of the same type and see that the configuration differences between two handlers of the same type resulted in the expected outcomes (i.e. differences in water use or tillage frequency). 

Once you have reviewed and updated the metadata as you think appropriate, this looks good to me!

File: input/metadata/properties/default.json (hunk 1 above)
Comment (Isgotkowitz) [reply input/metadata/properties/default.json]: Yes, good catch.

**Review by krb78 — APPROVED** [review]
I think it looks great, Ed! I ran it through and it works as it should :) Both for writing types that don't exist and making edits to config settings and resulting in changes in total manure output entering the reception pit. 🥳

File: RUFAS/routines/manure/IO_helpers/manure_manager_config_handler.py [hunk 2]
```diff
@@ -73,26 +74,34 @@ def get_custom_bedding_config(
...
         """
-        return self.custom_manure_handler_configs.get(
-            ManureHandlerType.get_type(manure_handler_type_name), None
-        )
+        try:
+            return self.manure_handler_configs[manure_handler_type_name]
+        except KeyError:
+            info_map = {"class": self.__class__.__name__, "function": self.get_manure_handler_config.__name__}
+            error_title = \
+                f"Attempted use a non-existent manure handler configuration called '{manure_handler_type_name}'."
```
Comment (Fjordrunner) [inline RUFAS/routines/manure/IO_helpers/manure_manager_config_handler.py]: I think this means to read "Attempted to use"?

File: RUFAS/routines/manure/manure_handlers/manure_handler_classes.py [hunk 3]
```diff
@@ -347,23 +297,23 @@ class ManureHandlerFactory:
     @classmethod
     def get_instance(
```
Comment (Fjordrunner) [inline RUFAS/routines/manure/manure_handlers/manure_handler_classes.py]: I would argue that the function name could be improved to either include information about the instance being a BaseManureHandler instance, and that it is a specific subtype of said class.

**Review by Fjordrunner — APPROVED** [review]
Looks good to me, but added a naming suggestion as well as pointed out a potential typo.

**Review by jadamchick — APPROVED** [review]
Looks great! logical and clearly organized

File: RUFAS/routines/manure/IO_helpers/manure_manager_config_handler.py (hunk 2 above)
Comment (Isgotkowitz) [reply RUFAS/routines/manure/IO_helpers/manure_manager_config_handler.py]: Nice catch

File: RUFAS/routines/manure/manure_handlers/manure_handler_classes.py (hunk 3 above)
Comment (Isgotkowitz) [reply RUFAS/routines/manure/manure_handlers/manure_handler_classes.py]: I have changed it to `get_manure_handler()`, and added all the children of `BaseManureHandler` as possible return types.

**Pull request merged**